Modify to return meaningful dictionary of available updates

theherk opened the pull request:
Previously list_updates was returning a formatted string including
tabs and newlines. This is no good for data consumtion and needed
to be changed to tabular data. We can now return a proper object
after consuming specific fields from the COM. Previously, the only
field included in the formatted return was description. Several
more, though not all, have been added.

This also adds a field parameter with which you can specify fields
for return. This parameter is superceded by the verbose parameter.

The documentation on the function has been updated to reflect
these changes.

The references to Harry Potter (as much as it pains me to mention
it) have been removed from the modified functions and methods.
The amount of those references in contrib was out of control. I'm
glad to see it making its way out.

Last, no changes have been made otherwise, so the download and
installation should not be affected.

theherk committed:
Modify to return meaningful dictionary of available updates

Previously list_updates was returning a formatted string including
tabs and newlines. This is no good for data consumtion and needed
to be changed to tabular data. We can now return a proper object
after consuming specific fields from the COM. Previously, the only
field included in the formatted return was description. Several
more, though not all, have been added.

This also adds a field parameter with which you can specify fields
for return. This parameter is superceded by the verbose parameter.

The documentation on the function has been updated to reflect
these changes.

The references to Harry Potter (as much as it pains me to mention
it) have been removed from the modified functions and methods.
The amount of those references in contrib was out of control. I'm
glad to see it making its way out.

Last, no changes have been made otherwise, so the download and
installation should not be affected.

theherk commented:
This previously would return:

```bash
pepper "herk*" --ignore-ssl-errors win_update.list_updates
{
    "return": [
        {
            "herk-hume-outer": "There are 6 updates, by category there are:\n\tUpdates: 5\n\tWindows 8.1: 5\n\tDefinition Updates: 1\n\tWindows Defender: 1\n"
        }
    ]
}
```

or:

```bash
pepper "herk*" --ignore-ssl-errors win_update.list_updates verbose=True
{
    "return": [
        {
            "herk-hume-outer": "There are 6 updates. They are as follows:\n\tUpdate for Windows 8.1 for x64-based Systems (KB3013816)\n\tUpdate for Windows 8.1 for x64-based Systems (KB3013769)\n\tUpdate for Windows 8.1 for x64-based Systems (KB3058168)\n\tUpdate for Windows 8.1 for x64-based Systems (KB3072019)\n\tUpdate for Windows 8.1 for x64-based Systems (KB3084905)\n\tDefinition Update for Windows Defender - KB2267602 (Definition 1.213.1492.0)\n"
        }
    ]
}
```

Yes, it printed prettier when returning in the default format:

```bash
salt "herk*" win_update.list_updates verbose=True
herk-hume-outer:
    There are 6 updates. They are as follows:
        Update for Windows 8.1 for x64-based Systems (KB3013816)
        Update for Windows 8.1 for x64-based Systems (KB3013769)
        Update for Windows 8.1 for x64-based Systems (KB3058168)
        Update for Windows 8.1 for x64-based Systems (KB3072019)
        Update for Windows 8.1 for x64-based Systems (KB3084905)
        Definition Update for Windows Defender - KB2267602 (Definition 1.213.1492.0)
```

That is beside the point. This is more useful as data than it is as a string. It now returns this:

```bash
pepper "herk*" --ignore-ssl-errors win_update_new.list_updates verbose=True
{
    "return": [
        {
            "herk-hume-outer": [
                {
                    "Categories": [
                        {
                            "Description": "A broadly released fix for a specific problem addressing a noncritical, non-security-related bug.",
                            "Name": "Updates"
                        },
                        {
                            "Description": "Windows 8.1",
                            "Name": "Windows 8.1"
                        }
                    ],
                    "Description": "Install this update to resolve issues in Windows. For a complete listing of the issues that are included in this update, see the associated Microsoft Knowledge Base article for more information. After you install this item, you may have to restart your computer.",
                    "Identity": {
                        "RevisionNumber": 201,
                        "UpdateID": "51962451-2048-4523-b9f5-3e2965f278de"
                    },
                    "IsMandatory": false,
                    "KBArticleIDs": [
                        "3013816"
                    ],
                    "MaxDownloadSize": "5074388",
                    "MinDownloadSize": "5074388",
                    "MoreInfoUrls": [
                        "http://support.microsoft.com/kb/3013816"
                    ],
                    "MsrcSeverity": "",
                    "ReleaseNotes": "",
                    "SecurityBulletinIDs": [],
                    "SupportUrl": "http://support.microsoft.com",
                    "Title": "Update for Windows 8.1 for x64-based Systems (KB3013816)"
                },
                {
                    "Categories": [
                        {
                            "Description": "A broadly released fix for a specific problem addressing a noncritical, non-security-related bug.",
                            "Name": "Updates"
                        },
                        {
                            "Description": "Windows 8.1",
                            "Name": "Windows 8.1"
                        }
                    ],
                    "Description": "Install this update to resolve issues in Windows. For a complete listing of the issues that are included in this update, see the associated Microsoft Knowledge Base article for more information. After you install this item, you may have to restart your computer.",
                    "Identity": {
                        "RevisionNumber": 209,
                        "UpdateID": "f80f971d-e079-4216-b686-bf5d4412376b"
                    },
                    "IsMandatory": false,
                    "KBArticleIDs": [
                        "3013769"
                    ],
                    "MaxDownloadSize": "58904256",
                    "MinDownloadSize": "58515593",
                    "MoreInfoUrls": [
                        "http://support.microsoft.com/kb/3013769"
                    ],
                    "MsrcSeverity": "",
                    "ReleaseNotes": "",
                    "SecurityBulletinIDs": [],
                    "SupportUrl": "http://support.microsoft.com",
                    "Title": "Update for Windows 8.1 for x64-based Systems (KB3013769)"
                }
            ]
        }
    ]
}
```

That is truncated for brevity. Also, you can request specific fields, or just get the titles.

```bash
pepper "herk*" --ignore-ssl-errors win_update_new.list_updates
{
    "return": [
        {
            "herk-hume-outer": [
                "Update for Windows 8.1 for x64-based Systems (KB3013816)",
                "Update for Windows 8.1 for x64-based Systems (KB3013769)",
                "Update for Windows 8.1 for x64-based Systems (KB3058168)",
                "Update for Windows 8.1 for x64-based Systems (KB3072019)",
                "Update for Windows 8.1 for x64-based Systems (KB3084905)",
                "Definition Update for Windows Defender - KB2267602 (Definition 1.213.1492.0)"
            ]
        }
    ]
}
```

theherk commented:
I see now that my build is failing due to not supporting Python 2.6. ><

I'll work on that soon.

cachedout commented:
@theherk Sounds good. Please let me know when you've fixed the 2.6 compatibility issues and we'll take a look. Thanks for your hard work on this one. (And yes, it's time to remove the Harry Potter references. It's a slightly sad day.) ;]

cachedout commented:
@theherk It also looks like this will no longer merge cleanly, so could you also please rebase it? Thanks.

theherk committed:
Modify dict comprehension to dict function with generator

This allows support < Python 2.7.

theherk commented:
@cachedout: I have corrected the compatibility issue and rebased. I am still having failures in Jenkins, and I am not sure why.

I get `test_issue_14979_output_file_permissions` in two of the current 3 failures. The third is a NullPointer Exception. I'm not sure what is causing the failures. If you could take a look, I would appreciate it.
.

cachedout commented:
Go Go Jenkins!

theherk commented:
Well there is one fewer failing now, so that's nice. Of the remaining 2 failures, the Jenkins jobs looks to consistently fail, so I don't think it is related to my code.

cachedout commented:
Thanks, @theherk . Merged.

theherk commented:
Fantastic. Thanks @cachedout. One last question. Since I will be deploying a lot of this in the future, I need to sort out if I'll be working from dev or if Boron is coming soon. Do you guys have an idea when Boron will be released?

cachedout commented:
@theherk Next couple of weeks for Boron.
